Kernel/Bus/VirtIO/Device.cpp

```diff
@@ -113,9 +127,11 @@ UNMAP_AFTER_INIT void Device::initialize()
             }
             config.offset = capability.read32(0x8);
             config.length = capability.read32(0xc);
-            // NOTE: Configuration length of zero is an invalid configuration that should be ignored.
-            if (config.length == 0)
-                continue;
+            // NOTE: Configuration length of zero is an invalid configuration.
+            // The VIRTIO_PCI_CAP_PCI_CFG configuration structure has length of 0
+            // but because we ignore that type and all other types should have a length
+            // greater than 0, we should assert in case this condition is not met.
+            VERIFY(config.length != 0);
```

Comment:
This looks like a "bad hardware can panic the kernel" check. I don't think we want this pattern right? What if it's copied somewhere where we can hotplug random devices? I would expect us to be tolerant of bad hardware and reject it rather than panicing.

Reply:
Mostly I agree with this, but this is VirtIO hardware so we can fully expect this length to be non-0. If the check fails it's either a new device we try to attach (which means, a developer is writing a new device driver) with a special configuration type we don't know about (which doesn't matter because the developer will need to implement a dozen more things so they will likely see kernel panics in other places), or someone touched this code and did something bad so we get zero-length. Hypervisors that implement VirtIO should technically always specify a valid length except for the VIRTIO_PCI_CAP_PCI_CFG apparently.

For hardware that is not VirtIO, handling edge cases like this is of course required. If people do a blind copy-and-paste of code snippets, we simply should not merge those patches.

I still need to investigate a little bit in the QEMU code so I can come up with a certain conclusion.